base_layer/wallet/src/ffi.rs

```diff
@@ -29,87 +29,402 @@
 //! will be exposed via FFI and will consist of API calls that the FFI client can make into the Wallet module and a set
 //! of Callbacks that the client must implement and provide to the Wallet module to receive asynchronous replies and
 //! updates.
-
-use crate::output_manager_service::service::PendingTransactionOutputs;
+extern crate libc;
+use crate::{output_manager_service::service::PendingTransactionOutputs, Wallet};
 use chrono::NaiveDateTime;
-use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
-use tari_comms::connection::NetAddress;
+use libc::{c_char, c_int, c_uchar, c_uint, c_ulonglong};
+// use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
+use crate::{
+    output_manager_service::{handle::OutputManagerResponse::TransactionCancelled, OutputManagerConfig},
+    wallet::WalletConfig,
+};
+use std::{
+    boxed::Box,
+    ffi::{CStr, CString},
+    process::Output,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+use tari_comms::{
+    connection::NetAddress,
+    control_service::ControlServiceConfig,
+    peer_manager::{NodeIdentity, Peer, PeerFeature, PeerFeatures, PeerNodeIdentity},
+};
+use tari_comms_dht::envelope::NodeDestination::NodeId;
 use tari_core::{
-    transaction::{Transaction, UnblindedOutput},
+    tari_amount::MicroTari,
+    transaction::{
+        OutputFeatures,
+        OutputFlags,
+        Transaction,
+        TransactionInput,
+        TransactionKernel,
+        TransactionOutput,
+        UnblindedOutput,
+    },
     types::{PrivateKey, PublicKey},
 };
+use tari_crypto::keys::SecretKey;
+use tari_p2p::initialization::CommsConfig;
+use tari_utilities::{hex::Hex, ByteArray};
+use tokio::runtime::Runtime;
 
-// use tari_core::transaction_protocol::sender::RawTransactionInfo;
+pub type TariWallet = Wallet;
+pub type WalletDateTime = NaiveDateTime;
 
-// TODO implement this
-pub struct WalletFfi {}
+/// -------------------------------- ByteVector ------------------------------------------------ ///
+pub struct ByteVector(Vec<c_uchar>); // declared like this so that it can be exposed to external header
 
-pub struct IdentityFfi {
-    public_key: PublicKey,
-    secret_key: PrivateKey,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_create() -> *mut ByteVector {
```

Comment:
We should use snake case. Not using a method to delimit words make reading the code difficult and is not in-line with the rest of the code base

Reply:
Is fixed for the next commit